src/engine/sparqlExpressions/SparqlExpressionValueGetters.cpp

```diff
@@ -252,7 +256,37 @@ std::optional<std::string> LanguageTagValueGetter::operator()(
     case Double:
     case Date:
     case Undefined:
-      return std::nullopt;
+      if constexpr (std::is_same_v<T, IdOrLiteralOrIri>) {
+        return Id::makeUndefined();
+      } else {
+        return std::nullopt;
+      }
   }
   AD_FAIL();
 }
+
+//_____________________________________________________________________________
+IdOrLiteralOrIri IriOrUriValueGetter::operator()(
+    ValueId id, const EvaluationContext* context) const {
+  return getValue<IdOrLiteralOrIri>(id, context, *this);
+}
+
+//______________________________________________________________________________
+std::optional<std::string> LanguageTagValueGetter::operator()(
+    ValueId id, const EvaluationContext* context) const {
+  return getValue<std::optional<std::string>>(id, context, *this);
+}
+
+//______________________________________________________________________________
+IdOrLiteralOrIri IriOrUriValueGetter::operator()(
+    const LiteralOrIri& litOrIri,
+    [[maybe_unused]] const EvaluationContext* context) const {
+  const auto& optIri = Iri::fromIrirefWithBasicUrlCheck(
+      litOrIri.isIri()
+          ? asStringViewUnsafe(litOrIri.getIri().getContent())
```

Comment:
If it already is an `IRI`, then you can directly return it without any check or conversion.